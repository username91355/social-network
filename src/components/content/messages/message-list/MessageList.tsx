import styles from "./MessageList.module.css";
import React, {ChangeEvent} from "react";
import Message from "./message/Message";
import {TMessage, TUser} from "../../../../data/reducers/messages-reducer";
import Button from "../../../common/button/Button";

type TProps = {
    dialogId: string
    dialogs: Array<TUser>
    messages: Array<TMessage>
    authUserPhoto: string
    authUserName: string
    newMessageText: string
    changeNewMessageArea: (value: string) => void
    sendMessage: (dialogID: number) => void
}

const MessageList: React.FC<TProps> = props => {

    const {
        dialogId,
        dialogs,
        messages,
        authUserPhoto,
        authUserName,
        newMessageText,
        changeNewMessageArea,
        sendMessage
    } = props

    const interlocutors = dialogs.filter((i: any) => (i.id === +dialogId))

    const textAreaOnChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
        changeNewMessageArea(e.currentTarget.value)
    }
    const onClickButtonsendMessage = () => {
        sendMessage(+dialogId)
    }

    if (messages.length === 0) {
        return <div className={styles.messages__messages_window_alt}>Select user for start dialog</div>
    }

    return (
        <div className={styles.messages__container}>
            <div className={styles.messages__messages_window}>
                {
                    messages.map((d,i) => {
                        return <Message key={i}
                                        outgoing={d.outgoing}
                                        message={d.message}
                                        authUserPhoto={authUserPhoto}
                                        authUserName={authUserName}
                                        interlocutors={interlocutors}/>
                    })
                }
            </div>
            <textarea className={styles.messages__textarea}
                      value={newMessageText}
                      onChange={textAreaOnChange}
                      placeholder={'Enter new message'}/>
            <div className={styles.messages__button_container}>
                <Button title={'Send'} onClick={onClickButtonsendMessage} />
            </div>
        </div>
    )
}

export default MessageList