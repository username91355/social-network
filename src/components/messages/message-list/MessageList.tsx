import styles from "./MessageList.module.css";
import {ChangeEvent} from "react";
import Message from "./message/Message";

const MessageList = (props: any) => {

    const {
        dialogId,
        dialogs,
        messages,
        authUserPhoto,
        newMessageText,
        changeNewMessageArea,
        sendMessage
    } = props

    const interlocutors = dialogs.filter((i: any) => (i.id === +dialogId))

    if (messages.length === 0) {
        return <div>Select user for start dialog</div>
    }

    return (
        <div className={styles.messages__container}>
            <div className={styles.messages__messages_window}>
                {
                    messages.map((d: any) => {
                        return <Message outgoing={d.outgoing}
                                        message={d.message}
                                        authUserPhoto={authUserPhoto}
                                        interlocutors={interlocutors}/>
                    })
                }
            </div>
            <hr/>
            <textarea className={styles.messages__textarea}
                      value={newMessageText}
                      onChange={(e: ChangeEvent<HTMLTextAreaElement>) =>
                          changeNewMessageArea(e.currentTarget.value)}
                      placeholder={'Enter new message'}/>
            <div className={styles.messages__button_container}>
                <button className={styles.messages__button_container_btn} onClick={() => sendMessage(dialogId)}>Send
                </button>
            </div>
        </div>
    )
}

export default MessageList