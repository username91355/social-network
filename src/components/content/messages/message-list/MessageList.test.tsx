import {fireEvent, render, screen} from "@testing-library/react";
import MessageList from "./MessageList";
import userEvent from "@testing-library/user-event";

const changeNewMessageArea = jest.fn()
const sendMessage = jest.fn()

const testData = {
    dialogId: '11117',
    dialogs: [{
        followed: true,
        id: 11117,
        name: 'Alice',
        photos: {
            small: 'smallAlice',
            large: 'largeAlice'
        },
        status: null,
        uniqueUrlName: null
    }],
    messages: [
        {id: 1, message: 'Hi. Thanks for the help.', outgoing: true},
        {id: 2, message: 'Hi. You are always welcome!', outgoing: false},
        {id: 3, message: 'How are you?', outgoing: false}
    ],
    authUserPhoto: 'authUserPhoto',
    authUserName: 'authUserName',
    newMessageText: 'newMessageText',
    changeNewMessageArea: changeNewMessageArea,
    sendMessage: sendMessage,
}

const testDataWithoutMessages = {
    dialogId: '11117',
    dialogs: [{
        followed: true,
        id: 11117,
        name: 'Alice',
        photos: {
            small: 'smallAlice',
            large: 'largeAlice'
        },
        status: null,
        uniqueUrlName: null
    }],
    messages: [],
    authUserPhoto: 'authUserPhoto',
    authUserName: 'authUserName',
    newMessageText: 'newMessageText',
    changeNewMessageArea: changeNewMessageArea,
    sendMessage: sendMessage,
}

describe('MessageList', () => {
    it('MessageList render', async () => {
        render(<MessageList {...testData}/>);

        const newMessageArea = screen.getByRole('textbox')
        const buttonSendMessage = screen.getByRole('button')

        expect(screen.getByText(/Hi. Thanks for the help./)).toBeInTheDocument()
        expect(screen.getByText(/Hi. You are always welcome!/)).toBeInTheDocument()
        expect(screen.getByText(/How are you?/)).toBeInTheDocument()
        expect(screen.getAllByText(/Alice/).length).toBe(2)

        fireEvent.change(newMessageArea, {target: { value: 'test' }})
        expect(testData.changeNewMessageArea).toHaveBeenCalledTimes(1)

        // Don`t work... input is controlled
        // await userEvent.type(newMessageArea, 'Test')
        // expect(newMessageArea).toHaveValue('Test')

        fireEvent.click(buttonSendMessage)
        expect(testData.sendMessage).toHaveBeenCalledTimes(1)

    });

    it('MessageList render without messages', async () => {
        render(<MessageList {...testDataWithoutMessages}/>);

        expect(screen.getByText(/Select user for start dialog/)).toBeInTheDocument()
    });

    it('MessageList snapshot', () => {
        const incomingMessage = render(<MessageList {...testData}/>);

        expect(incomingMessage).toMatchSnapshot()
    });
});

