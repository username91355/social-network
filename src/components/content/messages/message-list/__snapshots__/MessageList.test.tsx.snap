// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageList MessageList snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="messages__container"
      >
        <div
          class="messages__messages_window"
        >
          <div
            class="messages__outgoing"
          >
            <img
              alt="authUserPhoto"
              class="messages__outgoing_img"
              src="authUserPhoto"
            />
            <div
              class="messages__outgoing_message"
            >
              <h4>
                authUserName
              </h4>
              <span>
                Hi. Thanks for the help.
              </span>
            </div>
            <div />
          </div>
          <div
            class="messages__incoming"
          >
            <div />
            <div
              class="messages__incoming_message"
            >
              <h4>
                Alice
              </h4>
              <span>
                Hi. You are always welcome!
              </span>
            </div>
            <img
              alt="interlocutorsPhoto"
              class="messages__incoming_img"
              src="smallAlice"
            />
          </div>
          <div
            class="messages__incoming"
          >
            <div />
            <div
              class="messages__incoming_message"
            >
              <h4>
                Alice
              </h4>
              <span>
                How are you?
              </span>
            </div>
            <img
              alt="interlocutorsPhoto"
              class="messages__incoming_img"
              src="smallAlice"
            />
          </div>
        </div>
        <textarea
          class="messages__textarea"
          placeholder="Enter new message"
        >
          newMessageText
        </textarea>
        <div
          class="messages__button_container"
        >
          <button
            class="messages__button_container_btn"
          >
            Send
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="messages__container"
    >
      <div
        class="messages__messages_window"
      >
        <div
          class="messages__outgoing"
        >
          <img
            alt="authUserPhoto"
            class="messages__outgoing_img"
            src="authUserPhoto"
          />
          <div
            class="messages__outgoing_message"
          >
            <h4>
              authUserName
            </h4>
            <span>
              Hi. Thanks for the help.
            </span>
          </div>
          <div />
        </div>
        <div
          class="messages__incoming"
        >
          <div />
          <div
            class="messages__incoming_message"
          >
            <h4>
              Alice
            </h4>
            <span>
              Hi. You are always welcome!
            </span>
          </div>
          <img
            alt="interlocutorsPhoto"
            class="messages__incoming_img"
            src="smallAlice"
          />
        </div>
        <div
          class="messages__incoming"
        >
          <div />
          <div
            class="messages__incoming_message"
          >
            <h4>
              Alice
            </h4>
            <span>
              How are you?
            </span>
          </div>
          <img
            alt="interlocutorsPhoto"
            class="messages__incoming_img"
            src="smallAlice"
          />
        </div>
      </div>
      <textarea
        class="messages__textarea"
        placeholder="Enter new message"
      >
        newMessageText
      </textarea>
      <div
        class="messages__button_container"
      >
        <button
          class="messages__button_container_btn"
        >
          Send
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
